### PR TITLE
Detect extra space in keyword for better hint

### LINKED ISSUE #89640: Suggest joining two identifiers if the grammer is invalid and they would form a keyword
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9adebfa3cae4c99b1fa5c0c9f25e154a

```rust
fn main() {
    l et x = 5;
}
```

The current output is:

```
error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `et`
 --> src/main.rs:2:7
  |
2 |     l et x = 5;
  |       ^^ expected one of 8 possible tokens
```

Ideally the output should look like:

```
error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `et`
 --> src/main.rs:2:7
  |
2 |     l et x = 5;
  |       ^^ expected one of 8 possible tokens

     hint: did you accidentally put a space inside the keyword `let`?
```

Sometimes you accidentally insert whitespace inside a keyword, and might not notice it immediately. This minimal case is pretty obvious, but it might not actually always be that obvious if a lot is going on.

This could even be extended to cover all names in scope.

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #89640 

r? @Nilstrieb 

I met the same issue, then found out this old issue :)

### CODE REVIEW

File: tests/ui/parser/utf16-be-without-bom.stderr
Comment (chenyukang) [inline tests/ui/parser/utf16-be-without-bom.stderr:1]: Here is:
image

Comment (Nilstrieb) [reply tests/ui/parser/utf16-be-without-bom.stderr:1]: this isn't exactly great but I guess it's okay. I think finding such invalid tokens should actually be a fatal error, but I'm not sure about that. Maybe try putting up a PR that makes `source files must contain UTF-8 encoded text, unexpected null bytes might occur when a different encoding is used` errors fatal and see what wg-diagnostics thinks?

**Review by Noratrieb — APPROVED** [review]
r=me after fixing formatting

File: tests/ui/suggestions/issue-89640.stderr [hunk 1]
```diff
@@ -0,0 +1,13 @@
+error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `t`
+  --> $DIR/issue-89640.rs:2:8
+   |
+LL |     le t x: i32 = 3;
+   |        ^ expected one of 8 possible tokens
+   |
+help: did you accidentally put a space inside the keyword `let`
```
Comment (fmease) [inline tests/ui/suggestions/issue-89640.stderr:7]: The [suggestion style guide](https://rustc-dev-guide.rust-lang.org/diagnostics.html?highlight=illegal#suggestion-style-guide) advises against questions as suggestions. Not sure about the exact wording though, maybe:


```suggestion
help: you might have meant to write the keyword `let`
```

```suggestion
help: you might have meant to write the keyword `let` without the space
```


```suggestion
help: consider removed the space to spell keyword `let`
```

**Review by fmease — CHANGES_REQUESTED** [review]

File: tests/ui/suggestions/issue-89640.stderr [hunk 2]
```diff
@@ -0,0 +1,13 @@
+error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `t`
+  --> $DIR/issue-89640.rs:2:8
+   |
+LL |     le t x: i32 = 3;
+   |        ^ expected one of 8 possible tokens
+   |
+help: consider removed the space to spell keyword `let`
```
Comment (fmease) [inline tests/ui/suggestions/issue-89640.stderr:7]:
```suggestion
help: consider removing the space to spell keyword `let`
```

Sorry, I made a spelling error myself in the initial version of my suggestion. Thanks for applying my feedback! :)

**Review by fmease — CHANGES_REQUESTED** [review]

**Review by fmease — APPROVED** [review]
Ofc, the let-chain in `diagnostics.rs` is not formatted quite right (body indented by 8 spaces, `&&` indented by 2) but since `rustfmt` can't handle them yet it's fine. We're gonna have to format-the-world anyways once it does.

**Comment by chenyukang** [comment]
> Ofc, the let-chain in `diagnostics.rs` is not formatted quite right (body indented by 8 spaces, `&&` indented by 2) but since `rustfmt` can't handle them yet it's fine. We're gonna have to format-the-world anyways once it does.

yes, I was always mad with format when using `let-chains`, 
seems we already have a draft style for it: https://github.com/rust-lang/rust/pull/110568

**Comment by Nilstrieb** [comment]
Thank you GitHub for closing this PR!

**Comment by chenyukang** [comment]
> Thank you GitHub for closing this PR!

What happened? Seems it is a Github bug ....

**Comment by fmease** [comment]
Last time this happened to one of my PRs it got merged in the end with a one hour delay...

**Comment by chenyukang** [comment]
> Last time this happened to one of my PRs it got merged in the end with a one hour delay...

Github says this PR has conflicts:
image

While I checked there is no conflict...

**Comment by Nilstrieb** [comment]
I'll just close the PR manually. The commit is on master, GitHub is just confused. Everything is fine.

**Pull request closed without merging**